### ISSUE TITLE
Date.cweek doesn't support Date.beginning_of_week

### ISSUE THREAD

**Issue by markedmondson** [rails/rails#28610]
### Steps to reproduce
```ruby
> Date.beginning_of_week = :sunday
> Date.current.beginning_of_week.cweek
=> 12
> Date.current.end_of_week.cweek
=> 13
```

OR
```ruby
> Date.current.beginning_of_week(:sunday).cweek
=> 12
> Date.current.end_of_week(:sunday).cweek
=> 13
```

### Expected behavior
When the start of the week has been configured, it should modify the cweek calculation to correctly return the week number.

`strftime` behaves in the same way.

### Actual behavior
`Date.cweek` behaves as though the beginning_of_week hasn't been set.

### System configuration
**Rails version**: 4.2.8

**Ruby version**: 2.4.0

**Comment by yatindrarao** [comment]
Test case for this issue.

https://gist.github.com/yatindrarao/2aa0bca1a32e708264636cfc62d6451a

**Comment by pixeltrix** [comment]
Looking at the Ruby documentation for the [calendar week date][1] format it's clear that it's implementing the [ISO 8601 week date][2] format which defines the start of the week as Monday.

From the Ruby docs:

    The calendar week is a seven day period within a calendar year, starting on a
    Monday and identified by its ordinal number within the year; the first calendar
    week of the year is the one that includes the first Thursday of that year. In the
    Gregorian calendar, this is equivalent to the week which includes January 4.

From the Wikipedia page

    Weeks start with Monday. Each week's year is the Gregorian year in which the
    Thursday falls. The first week of the year, hence, always contains 4 January.
    ISO week year numbering therefore slightly deviates from the Gregorian for
    some days close to 1 January.

The ISO 8601 standard doesn't allow for weeks starting on any other day so it's working as expected.

Additionally, `Date#cweek` is a Ruby native method and `Date.beginning_of_the_week` is a Rails config option and we prefer not patching native methods and provide alternative APIs instead, e.g `Date.today` is native, whereas `Date.current` is Rails.

If you can provide a real world use case (e.g. a specific country where the week starts on a different day and has different week numbering to ISO 8601) then we could look at adding an additional API method.

[1]: https://ruby-doc.org/stdlib-2.4.1/libdoc/date/rdoc/Date.html#class-Date-label-Week+Date
[2]: https://en.wikipedia.org/wiki/ISO_week_date

**Comment by markedmondson** [comment]
A fair argument. My reasoning is more of expectation than the specificity of the docs with regard to ISO 8601.

Another thing that drew me to questioning the implementation is the [`Date` docs for strftime](https://ruby-doc.org/stdlib-2.4.1/libdoc/date/rdoc/Date.html#method-i-strftime) which have implementations for ISO 8601 **and** weeks beginning with Sunday (see the 2nd part).

> ISO 8601 week-based year and week number:
> The week 1 of YYYY starts with a Monday and includes YYYY-01-04.
> The days in the year before the first week are in the last week of
the previous year.
```
  %G - The week-based year
  %g - The last 2 digits of the week-based year (00..99)
  %V - Week number of the week-based year (01..53)
```

> Week number:
> The week 1 of YYYY starts with a Sunday or Monday (according to %U
or %W).  The days in the year before the first week are in week 0.
```
  %U - Week number of the year.  The week starts with Sunday.  (00..53)
  %W - Week number of the year.  The week starts with Monday.  (00..53)
```

Now the use of setting `Date.beginning_of_week` and using `strftime "%U"` adds more confusion:
```
> Date.beginning_of_week = :sunday
=> :sunday
> Date.current.beginning_of_week.strftime "%U"
=> "13"
> Date.current.end_of_week.strftime "%U"
=> "13"
> Date.beginning_of_week = :saturday
=> :saturday
> Date.current.beginning_of_week.strftime "%U"
=> "12"
> Date.current.end_of_week.strftime "%U"
=> "13"
```

To your second point, while ISO 8601 is the standard, when utilizing `Date` for building a calendar, users anticipate seeing a calendar starting with the beginning of the week as per their notion of a week, maybe think of it like metric vs imperial. There's [some further Wikipedia reference](https://en.wikipedia.org/wiki/Week#Week_numbering) regarding different regions with differing beginnings of week (granted, there is related discussion regarding what is actually considered fact).

To summarize, I'm not disputing that if implementing ISO 8601 date (as the ruby `Date` class does) anything is incorrect, moreso that perhaps the implementation of adding `Date.beginning_of_week=` doesn't correctly address all the possible associated cases and causes potential confusion.

**Comment by markedmondson** [comment]
@pixeltrix wanted to call your attention to my response since this is closed I wasn't sure if you saw it.
